src/discovery/Discovery.ts

```diff
@@ -123,36 +158,34 @@ class Discovery {
   } = {}): Promise<void> {
     this.logger.info(`Starting ${this.constructor.name}`);
     if (fresh) {
-      await this.db.clear(this.discoveryDbPath);
-    }
-    // Getting latest ID and creating ID generator
-    let latestId: DiscoveryQueueId | undefined;
-    const keyIterator = this.db.iterator(this.discoveryQueueDbPath, {
-      limit: 1,
-      reverse: true,
-      values: false,
-    });
-    for await (const [keyPath] of keyIterator) {
-      const key = keyPath[0] as Buffer;
-      latestId = IdInternal.fromBuffer<DiscoveryQueueId>(key);
+      // Cancel all tasks for discovery
+      for await (const task of this.taskManager.getTasks('asc', true, [
+        this.constructor.name,
+      ])) {
+        task.cancel(discoveryDestroyedTaskReason);
+      }
     }
-    this.discoveryQueueIdGenerator =
-      discoveryUtils.createDiscoveryQueueIdGenerator(latestId);
-    // Starting the queue
-    this.discoveryProcess = this.setupDiscoveryQueue();
+    this.taskManager.registerHandler(
+      this.discoverVertexHandlerId,
+      this.discoverVertexHandler,
+    );
     this.logger.info(`Started ${this.constructor.name}`);
   }
 
   public async stop(): Promise<void> {
     this.logger.info(`Stopping ${this.constructor.name}`);
-    this.queuePlug.resolveP();
-    await this.discoveryProcess;
+    this.taskManager.deregisterHandler(this.discoverVertexHandlerId);
     this.logger.info(`Stopped ${this.constructor.name}`);
   }
```

Comment:
It is a good idea to also initiate cancellation of any discovery tasks that still exist. We should not necessarily expect that the tasks have been stopped when calling `Discovery.stop`. This would then make this similar to `NodeManager.stop` which does in fact remove all tasks.

In this case, we don't need to remove all tasks, but just send cancellation signals and then wait for them to resolve/fulfill.

We also should check if `TaskManager` is still processing and if it is, we need to throw an exception. The `NodeManager` should probably be checking for this too.

Reply:
`Discovery` is now cancelling any active tasks when stopping. Both `NodeManager` and `Discovery` check for if `TaskManager` is still processing when stopping.